### ISSUE TITLE
$.Rule in TOKEN() gives error.

### ISSUE THREAD

**Issue by blinker29** [tree-sitter/tree-sitter#449]
It looks like any rule reference within a TOKEN() will produce an error.

```
rules:
{
  A: $ => token($.B),
  B: $ => seq("a", "b")
}
Error processing rule A_token1
Details:
  Grammar error: Unexpected rule Symbol(Symbol { kind: NonTerminal, index: 1 })
```

This looks like a bug, but I'm new to TS.  If user error, please explain.

**Comment by maxbrunsfeld** [comment]
This is the (current) intended behavior; the point of `token` is to indicate that the given rule should be parsed as one token (as opposed to a parent node with children), whereas symbol references correspond to parent-child relationships in the syntax tree.

It would be possible to relax this restriction by automatically 'inlining' the content of the children into the parent, to form one single token. Maybe we should do that.

**Comment by blinker29** [comment]
Okay, thank you.  No idea of the effort involved, but expanding use cases for token() is akin to implementing the "$" operator in PEGJS.  Promotes reuse/encapsulation/clarity.  On the other hand, if all token() expressions have to be terminals, then there will be more duplication of expressions and fewer mnemonics, which reduces readability, maintainability, etc.

**Comment by bookofproofs** [comment]
> Maybe we should do that.

Yes, please! It would further increase the readability of the AST and how intuitively we can interpret it.

**Comment by Hubro** [comment]
I would like to chime in in favor of lifting this restriction. I'm trying to implement a parser for YANG, and this restriction is making my grammar much less readable.

Context: https://github.com/tree-sitter/tree-sitter/discussions/1266

**Comment by ahlinc** [comment]
**Future implementation note:**
- It makes sense to implement optimization on the generator level that if some `$.Rule` used only as inlined into tokens then it makes sense to remove it from a regular rules set.

**Comment by MeepTech** [comment]
Any progress on this? This kind of makes tree-sitter unusable for any syntax that uses prefixes or suffixes as something like:
```
{  _prefixed_expression: $ => seq($.prefix, token.immediate($.expression)) }
```

doesn't work and requires a rather frustrating destruction of the entire language syntax in a rather inelegant way; As @Hubro described.

**Comment by maxbrunsfeld** [comment]
Just to clarify - You have always been able to use normal variables to extract whatever you want, since the grammars are JavaScript. So it’s a somewhat minor stylistic change that this feature would enable.
